feat(limit-count): add sliding window support for redis-based policies

sihyeonn opened the pull request:
### Description

This PR introduces **sliding window** and **approximate sliding window** support to the limit-count plugin for Redis- and Redis Cluster–based policies, while preserving the existing fixed window behavior by default.

#### Key Changes

1. **New `window_type` option**: Adds a configurable `window_type` parameter with three modes:
   - `"fixed"` (default): Traditional fixed window rate limiting
   - `"sliding"`: Exact sliding window using Redis sorted sets for precise per-request tracking
   - `"approximate_sliding"`: Memory-efficient approximate sliding window using counter-based algorithm

2. **Policy restriction**: Enforces that `window_type` other than `"fixed"` is only allowed when the policy is `redis` or `redis-cluster` to avoid unsupported combinations.

3. **Implementation**:
   - **Exact sliding window** (`"sliding"`): Uses Redis sorted sets (ZADD, ZCARD, ZREMRANGEBYSCORE) to track individual requests within the time window, providing 100% accuracy but higher memory usage (~100 bytes per request).
   - **Approximate sliding window** (`"approximate_sliding"`): Based on [Cloudflare's algorithm](https://blog.cloudflare.com/counting-things-a-lot-of-different-things/), uses two window counters with weighted calculation. Provides ~99.997% accuracy (tested with 400B+ requests) with significantly lower memory footprint (~16 bytes per key).

4. **Test coverage**: Adds comprehensive test cases for both sliding window modes in Redis and Redis Cluster configurations:
   - `t/plugin/limit-count-redis-sliding.t`
   - `t/plugin/limit-count-redis-cluster-sliding.t`
   - `t/plugin/limit-count-redis-approximate-sliding.t`
   - `t/plugin/limit-count-redis-cluster-approximate-sliding.t`

5. **Documentation**: Updates English and Chinese limit-count plugin documentation to describe the new `window_type` option, behavior differences, and performance/accuracy trade-offs.

#### Backward Compatibility

The default behavior remains unchanged for existing configurations. Configurations that do not specify `window_type` or explicitly use `"fixed"` will continue to work as before.

#### Performance Considerations

- **Exact sliding window**: Higher accuracy, higher memory usage, suitable for strict rate limiting requirements
- **Approximate sliding window**: Near-perfect accuracy with minimal memory overhead, suitable for high-traffic scenarios where slight approximation is acceptable

### Which issue(s) this PR fixes:
Fixes #

### Checklist
- [x] I have explained the need for this PR and the problem it solves
- [x] I have explained the changes or the new features added to this PR
- [x] I have added tests corresponding to this change
- [x] I have updated the documentation to reflect this change
- [x] I have verified that this change is backward compatible

sihyeonn committed:
feat: add window_type for limit-count

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn committed:
feat: add limit-count sliding window for redis with tests

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn committed:
feat: add limit-count sliding window for redis-cluster with tests

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn committed:
chore: update docs about sliding window

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

Baoyuantop commented on apisix/plugins/limit-count/limit-count-redis.lua:
```diff
@@ -40,12 +40,61 @@ local script = core.string.compress_script([=[
+    end
+
+    for i = 1, cost do
+        redis.call('ZADD', KEYS[1], now, now .. ':' .. i)
```
In high-concurrency scenarios, uniqueness may not be guaranteed. I think we can utilize ngx.var.request_id

sihyeonn replied:
Sorry to late and also thanks for the review and the suggestion about using ngx.var.request_id for uniqueness under high concurrency.
I’ve updated both Redis and Redis Cluster sliding window implementations so that the Lua scripts now always use ngx.var.request_id as part of the ZSET member value (i.e. "$request_id:<i>").

sihyeonn committed:
feat: handling the case of same timestamp

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn committed:
chore: add tests for handling same timestamp

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn commented on apisix/plugins/limit-count/limit-count-redis-cluster.lua:
```diff
@@ -39,7 +40,56 @@ local script = core.string.compress_script([=[
+    for i = 1, cost do
+        local member = req_id .. ':' .. i
+        redis.call('ZADD', KEYS[1], now, member)
+    end
```
@nic-6443 

Thanks for raising this concern!

In this sliding-window implementation, we intentionally trade some Redis work per request for *exact* “N requests per rolling window” semantics. However, the memory and per-key load are bounded:

- on every request we first call `ZREMRANGEBYSCORE` to drop all events outside of the current window and then enforce `current + cost <= limit` before inserting new members,
- rejected requests do **not** insert new entries,
- so for each key the ZSET cardinality is always `<= limit`, and Redis memory for this key is bounded by that limit.

Because of this, even though we do one `ZADD` per allowed request, we don’t end up with unbounded growth like in some TTL-only designs – the sliding window state per key stays O(limit). Also, `window_type = "sliding"` is an opt-in mode and the docs explicitly call out that it is more expensive than the default fixed window, so it’s intended for use cases that really need accurate rolling windows and can accept the extra Redis cost.

nic-6443 replied:
I suggest you use the approximate sliding window mentioned in this blog https://blog.cloudflare.com/counting-things-a-lot-of-different-things/ to implement this function. You can refer to the code here: https://github.com/ElvinEfendi/lua-resty-global-throttle/blob/main/lib/resty/global_throttle/sliding_window.lua

sihyeonn replied:
Sorry for the late response! Pushed a fix addressing the remaining feedback — CROSSSLOT prevention with hash tags, TTL unit normalization (ms to seconds), request_id fallback, docs for approximate_sliding, and test improvements. Please take a look when you get a chance.

sihyeonn committed:
feat(limit-count): add approximate sliding window option

Add `window_type = "approximate_sliding"` option that implements
the approximate sliding window algorithm (as described in Cloudflare's
blog) alongside the existing exact sliding window implementation.

Users can now choose between three window types:
- "fixed": Fixed window algorithm (default)
- "sliding": Exact sliding window using sorted sets
- "approximate_sliding": Approximate sliding window using counters

The approximate algorithm provides significant performance benefits:
- Memory: ~16 bytes per key vs ~100 bytes per request
- Operations: Simple INCR vs ZADD + ZREMRANGEBYSCORE + ZCARD
- Accuracy: 0.003% error rate (based on 400B+ requests analysis)

Changes:
- Add "approximate_sliding" to window_type enum in schema
- Implement approximate sliding algorithm in Redis and Redis Cluster
- Add test cases for approximate sliding window behavior
- Optimize INCR operations and add reset value validation

Reference: https://blog.cloudflare.com/counting-things-a-lot-of-different-things/

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

moonming reviewed (changes requested):
Hi @sihyeonn, thank you for the sliding window rate limiting implementation!

At 1816 lines, this is a significant feature addition. Sliding window is more accurate than fixed window for rate limiting, and using Redis sorted sets is a well-known approach.

**Key concerns**:
1. **Redis overhead**: Each request requires multiple Redis commands (ZADD, ZREMRANGEBYSCORE, ZCARD, EXPIRE). Have you benchmarked the latency impact compared to the current fixed window approach?
2. **Memory usage**: Sorted sets store individual timestamps, which can consume significant Redis memory under high traffic. What's the memory footprint for, say, 10K requests/second with a 60-second window?
3. **Atomicity**: Are the Redis operations wrapped in a pipeline or Lua script to ensure atomicity?
4. **Migration path**: How do existing users migrate from fixed window to sliding window? Is it a simple config change?

Please provide performance benchmarks comparing the two approaches. This will help us make an informed decision about including it. Thank you for the thorough implementation!

sihyeonn committed:
fix(limit-count): address PR review feedback for sliding window

- Fix assert in script_sliding and script_approximate_sliding to validate
  ARGV[4] (cost) instead of ARGV[3] (limit) in both redis and redis-cluster
- Add fallback for ngx.var.request_id when it may not be available
- Add approximate_sliding to window_type enum in EN and ZH docs
- Use Redis hash tags {key}:suffix in redis-cluster approximate_sliding
  script to prevent CROSSSLOT errors
- Move redis-cluster tests from redis-sliding test files to their
  dedicated cluster-specific test files

sihyeonn committed:
fix: address review feedback for sliding window support

- Use hash tags ({key}:window_id) in approximate_sliding Lua script
  for redis.lua to avoid potential CROSSSLOT errors
- Convert sliding/approximate_sliding reset values from ms to seconds
  so X-RateLimit-Reset header returns correct values
- Add worker ID to request_id fallback for better uniqueness when
  ngx.var.request_id is nil
- Document approximate_sliding mode in Description section of both
  English and Chinese docs
- Add Redis Cluster availability check (skip_all) to cluster sliding
  window tests so they skip gracefully when cluster is unavailable
- Replace hardcoded 404 status with \d+ pattern in test assertions
  for response_body_like sections

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn commented on docs/en/latest/plugins/limit-count.md:
```diff
@@ -46,6 +51,7 @@ You may see the following rate limiting headers in the response:
 | ----------------------- | ------- | ----------------------------------------- | ------------- | -------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
 | count                   | integer | True                                      |               | > 0                              | The maximum number of requests allowed within a given time interval.                                                                                                                                                                                                       ...
 | time_window             | integer | True                                      |               | > 0                        | The time interval corresponding to the rate limiting `count` in seconds.                                                                                                                                                                                                         ...
+| window_type             | string  | False                                     | fixed         | ["fixed","sliding"]          | The window behavior type. `fixed` uses a fixed window algorithm. `sliding` uses a sliding window algorithm to enforce an exact number of requests per rolling window. `sliding` is only supported when `policy` is `redis` or `redis-cluster`.                                 ...
```
Added — both English and Chinese docs now include approximate_sliding with a description.

sihyeonn commented on t/plugin/limit-count-redis-sliding.t:
```diff
@@ -0,0 +1,501 @@
+            local t = require("lib.test_admin").test
+            -- Note: This test requires redis-cluster to be available
+            -- It validates schema but may fail at runtime if cluster unavailable
+            local code, body = t('/apisix/admin/routes/6',
```
Moved Redis Cluster test cases behind a skip_all check that verifies cluster availability before running.

sihyeonn committed:
fix(limit-count): resolve luacheck warnings in sliding window implementation

- Declare local err alongside res to avoid undefined variable warnings
- Extract worker_id to a local variable to stay under 100-char line limit
- Break long string concatenation in error message across two lines

Signed-off-by: Sihyeon Jang <sihyeon.jang@navercorp.com>

sihyeonn committed:
Merge remote-tracking branch 'upstream/master' into feat/sh-sliding-window

# Conflicts:
#	apisix/plugins/limit-count/limit-count-redis-cluster.lua
#	apisix/plugins/limit-count/limit-count-redis.lua
